### PR TITLE
initial android fixes

### PR DESCRIPTION
**Pull request by headupinclouds**
For discussion...

For a minimal patch philosophy we can define `NSYNC_LANGUAGE=c++11` externally (most cases now) to leverage existing CMakeLists.txt logic checks:

```
polly.py --toolchain android-ndk-r10e-api-19-armeabi-v7a-neon --install --reconfig --fwd NSYNC_LANGUAGE=c++11
```

`SUCCESS`

It might be better to use built-in CMake language checks.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -100,7 +103,16 @@ if ("${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
 		)
 		set (NSYNC_TEST_OS_SRC
 			"platform/posix/src/start_thread.c"
+        )
+    elseif ( "${CMAKE_SYSTEM_NAME}X" STREQUAL "AndroidX")
```
Comment (ruslo) [inline CMakeLists.txt]: `elseif(ANDROID)` should work fine

File: CMakeLists.txt [hunk 2]
```diff
@@ -100,7 +103,16 @@ if ("${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
...
+		include_directories ("${PROJECT_SOURCE_DIR}/platform/posix")
+		add_compile_options ("-std=c++11")
```
Comment (ruslo) [inline CMakeLists.txt]: Belongs to toolchain, no need to set it here.

File: CMakeLists.txt [hunk 3]
```diff
@@ -155,6 +167,26 @@ if (NOT "${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
 	elseif ("${CMAKE_SYSTEM_NAME}X" STREQUAL "OpenBSDX")
 		include_directories ("${PROJECT_SOURCE_DIR}/platform/openbsd")
 		set (NSYNC_POSIX ON)
+      elseif ( "${CMAKE_SYSTEM_NAME}X" STREQUAL "AndroidX")
```
Comment (ruslo) [inline CMakeLists.txt]: `if(ANDROID)` can be used

File: CMakeLists.txt [hunk 4]
```diff
@@ -155,6 +167,26 @@ if (NOT "${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
...
+
+        message("ANDROID!!!")
+        
+      if ("${ANDROID_ARCH}X" STREQUAL "armv7-aX")
```
Comment (ruslo) [inline CMakeLists.txt]: There is no `ANROID_ARCH` variable. Available variables to analyze:
* https://cmake.org/cmake/help/latest/variable/CMAKE_ANDROID_ARCH.html
* https://cmake.org/cmake/help/latest/variable/CMAKE_ANDROID_ARCH_ABI.html

File: CMakeLists.txt [hunk 5]
```diff
@@ -155,6 +167,26 @@ if (NOT "${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
...
+		include_directories ("${PROJECT_SOURCE_DIR}/platform/arm")
+		set (NSYNC_POSIX ON)        
+      elseif ("${ANDROID_ARCH}X" STREQUAL "armv7-a-hardX")
```
Comment (ruslo) [inline CMakeLists.txt]: `...X` trick is not saving us from anything, it's better to use `string(COMPARE EQUAL ...)` or update requirement to CMake 3.1+.

**Review by ruslo — CHANGES_REQUESTED** [review]

**Comment by ruslo** [comment]
> It might be better to use built-in CMake language checks

It theory we can use [CMAKE_CXX_STANDARD](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html) but need to update Polly:
* https://github.com/ruslo/polly/blob/c908e54dbb23cbc327930b14ef31a8cd5af064c1/flags/cxx11.cmake#L14-L19

File: CMakeLists.txt (hunk 3 above)
Comment (headupinclouds) [reply CMakeLists.txt]: Done.

File: CMakeLists.txt (hunk 2 above)
Comment (headupinclouds) [reply CMakeLists.txt]: I agree in theory.  In this case I was aiming for a minimal patch (replicating existing syntax/logic).    Do you prefer updating the rest of the CMakeLists.txt?

File: CMakeLists.txt (hunk 4 above)
Comment (headupinclouds) [reply CMakeLists.txt]: Done.

File: CMakeLists.txt (hunk 2 above)
Comment (ruslo) [reply CMakeLists.txt]: > Do you prefer updating the rest of the CMakeLists.txt?

No, minimal patch seems fine. Just no need to add new line with `add_compile_options ("-std=c++11")`.

File: CMakeLists.txt (hunk 1 above)
Comment (headupinclouds) [reply CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 5 above)
Comment (headupinclouds) [reply CMakeLists.txt]: I made use of `string(COMARE EQUAL ...)` for the localized android specific patches.

File: CMakeLists.txt [hunk 6]
```diff
@@ -155,6 +167,26 @@ if (NOT "${NSYNC_LANGUAGE}X" STREQUAL "c++11X")
 	elseif ("${CMAKE_SYSTEM_NAME}X" STREQUAL "OpenBSDX")
 		include_directories ("${PROJECT_SOURCE_DIR}/platform/openbsd")
 		set (NSYNC_POSIX ON)
+    elseif (ANDROID)
+      # Run string comparisons non-lazily to avoid excessive nesting
+      string(COMPARE EQUAL "${CMAKE_ANDROID_ARCH}" "armv7-a" _armv7_a)
```
Comment (ruslo) [inline CMakeLists.txt]: `CMAKE_ANDROID_ARCH` can't have value `armv7-a`. Possible values listed here:
* https://cmake.org/cmake/help/latest/variable/CMAKE_ANDROID_ARCH.html

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (headupinclouds) [reply CMakeLists.txt]: Ok.  For now I'll leave it as is?  Let me know if you still think a change is required in this PR.

Comment (ruslo) [reply CMakeLists.txt]: > For now I'll leave it as is?

Ok

File: CMakeLists.txt (hunk 6 above)
Comment (headupinclouds) [reply CMakeLists.txt]: I see.  I was reading from the `ANDROID_ARCH` *property* documentation.  Thanks for the correction.

https://cmake.org/cmake/help/latest/prop_tgt/ANDROID_ARCH.html?highlight=android_arch#android-arch

Comment (headupinclouds) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt [hunk 7]
```diff
@@ -28,6 +28,9 @@ set (NSYNC_POSIX_SRC
 # is false when ${CMAKE_C_COMPILER_ID} has the value "MSVC"!  See
 #    https://cmake.org/cmake/help/v3.1/policy/CMP0054.html
 
+message("NSYNC_LANGUAGE = ${NSYNC_LAGUAGE}")
+message("CMAKE_SYSTEM_NAME = ${CMAKE_SYSTEM_NAME}")
```
Comment (ruslo) [inline CMakeLists.txt]: Remove?

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 7 above)
Comment (headupinclouds) [reply CMakeLists.txt]: Removed!

**Review by ruslo — APPROVED** [review]

**Review by ruslo — APPROVED** [review]
Looks good 👍

**Comment by headupinclouds** [comment]
Travis is still churning away.  I guess we'll find out about ios next?  

{Re: NSYNC_LANGUAGE=c++11}: I suppose we can `--fwd` this manually for the CI toolchains that need it.

**Comment by headupinclouds** [comment]
We can make a list of the various mini `google/<proj>` dependencies that we're adding as part of the epic `add tensorflow` issue.  Since these are all typically atomic packages (no `HunterGate()`), most of the CMakeLists.txt patches (package config installation and/or adding CMakeLists.txt)  can probably be submitted upstream.  Tensorflow is a great use case for Hunter, actually.

**Comment by headupinclouds** [comment]
I'll leave this open while the travis test is still running.

**Comment by ruslo** [comment]
> I suppose we can --fwd this manually for the CI toolchains that need it

We need to think about packages that will use it as dependency.

Solutions:
* Update Polly: use `CMAKE_CXX_STANDARD` instead of `-std=c++11`. In this case we can analyze `CMAKE_CXX_STANDARD`. In theory this should "just work" seamlessly but in practice we may break some custom build scheme packages that don't understand `CMAKE_CXX_STANDARD` (though it's a good reason to fix them).
* Use `hunter_check_toolchain_definition` in `hunter.cmake` to detect C++11 and add `NSYNC_LANGUAGE` by `CMAKE_ARGS`: https://docs.hunter.sh/en/latest/reference/user-modules/hunter_check_toolchain_definition.html

**Comment by headupinclouds** [comment]
Merging this initial/minimal patch.  I'll migrate the CMake/C++11 related TODO to the issues for tracking.

UPDATE: Issues need to explicitly enabled here.

**Comment by ruslo** [comment]
> Issues need to explicitly enabled here

Done: https://github.com/hunter-packages/nsync/issues

**Pull request merged**